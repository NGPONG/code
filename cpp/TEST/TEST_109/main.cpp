#include <iostream>
#include <thread>
#include <memory>
#include <chrono>
#include <csignal>
#include <dlfcn.h>

using namespace std::chrono_literals;

extern std::int32_t x;
extern std::int32_t z;

extern "C" void foo();

int main (int argc, char *argv[]) {
  std::raise(SIGSTOP);

  std::cout << x << std::endl;
  std::cout << z << std::endl;

  foo();

  // =================================================================

  typedef void (* foo_fp)(void);

  void *handler = dlopen("./libvar.so", RTLD_LOCAL | RTLD_NOW);

  foo_fp handler_foo = (foo_fp)dlsym(handler, "foo");
  int *handler_x = (int *)dlsym(handler, "x");
  int *handler_z = (int *)dlsym(handler, "z");

  std::cout << *handler_x << std::endl;
  std::cout << *handler_z << std::endl;

  handler_foo();

  return 0;
}
