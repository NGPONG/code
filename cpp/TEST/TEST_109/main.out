
main:     file format elf64-x86-64


Disassembly of section .init:

0000000000001000 <_init>:
    1000:	f3 0f 1e fa          	endbr64 
    1004:	48 83 ec 08          	sub    $0x8,%rsp
    1008:	48 8b 05 d9 2f 00 00 	mov    0x2fd9(%rip),%rax        # 3fe8 <__gmon_start__@Base>
    100f:	48 85 c0             	test   %rax,%rax
    1012:	74 02                	je     1016 <_init+0x16>
    1014:	ff d0                	call   *%rax
    1016:	48 83 c4 08          	add    $0x8,%rsp
    101a:	c3                   	ret    

Disassembly of section .plt:

0000000000001020 <.plt>:
    1020:	ff 35 5a 2f 00 00    	push   0x2f5a(%rip)        # 3f80 <_GLOBAL_OFFSET_TABLE_+0x8>
    1026:	f2 ff 25 5b 2f 00 00 	bnd jmp *0x2f5b(%rip)        # 3f88 <_GLOBAL_OFFSET_TABLE_+0x10>
    102d:	0f 1f 00             	nopl   (%rax)
    1030:	f3 0f 1e fa          	endbr64 
    1034:	68 00 00 00 00       	push   $0x0
    1039:	f2 e9 e1 ff ff ff    	bnd jmp 1020 <_init+0x20>
    103f:	90                   	nop
    1040:	f3 0f 1e fa          	endbr64 
    1044:	68 01 00 00 00       	push   $0x1
    1049:	f2 e9 d1 ff ff ff    	bnd jmp 1020 <_init+0x20>
    104f:	90                   	nop
    1050:	f3 0f 1e fa          	endbr64 
    1054:	68 02 00 00 00       	push   $0x2
    1059:	f2 e9 c1 ff ff ff    	bnd jmp 1020 <_init+0x20>
    105f:	90                   	nop
    1060:	f3 0f 1e fa          	endbr64 
    1064:	68 03 00 00 00       	push   $0x3
    1069:	f2 e9 b1 ff ff ff    	bnd jmp 1020 <_init+0x20>
    106f:	90                   	nop
    1070:	f3 0f 1e fa          	endbr64 
    1074:	68 04 00 00 00       	push   $0x4
    1079:	f2 e9 a1 ff ff ff    	bnd jmp 1020 <_init+0x20>
    107f:	90                   	nop

Disassembly of section .plt.got:

0000000000001080 <__cxa_finalize@plt>:
    1080:	f3 0f 1e fa          	endbr64 
    1084:	f2 ff 25 2d 2f 00 00 	bnd jmp *0x2f2d(%rip)        # 3fb8 <__cxa_finalize@GLIBC_2.2.5>
    108b:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

Disassembly of section .plt.sec:

0000000000001090 <__cxa_atexit@plt>:
    1090:	f3 0f 1e fa          	endbr64 
    1094:	f2 ff 25 f5 2e 00 00 	bnd jmp *0x2ef5(%rip)        # 3f90 <__cxa_atexit@GLIBC_2.2.5>
    109b:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

00000000000010a0 <raise@plt>:
    10a0:	f3 0f 1e fa          	endbr64 
    10a4:	f2 ff 25 ed 2e 00 00 	bnd jmp *0x2eed(%rip)        # 3f98 <raise@GLIBC_2.2.5>
    10ab:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

00000000000010b0 <_Z3foov@plt>:
    10b0:	f3 0f 1e fa          	endbr64 
    10b4:	f2 ff 25 e5 2e 00 00 	bnd jmp *0x2ee5(%rip)        # 3fa0 <_Z3foov@Base>
    10bb:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

00000000000010c0 <_ZNSt8ios_base4InitC1Ev@plt>:
    10c0:	f3 0f 1e fa          	endbr64 
    10c4:	f2 ff 25 dd 2e 00 00 	bnd jmp *0x2edd(%rip)        # 3fa8 <_ZNSt8ios_base4InitC1Ev@GLIBCXX_3.4>
    10cb:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

00000000000010d0 <_ZNSolsEi@plt>:
    10d0:	f3 0f 1e fa          	endbr64 
    10d4:	f2 ff 25 d5 2e 00 00 	bnd jmp *0x2ed5(%rip)        # 3fb0 <_ZNSolsEi@GLIBCXX_3.4>
    10db:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

Disassembly of section .text:

00000000000010e0 <_start>:
    10e0:	f3 0f 1e fa          	endbr64 
    10e4:	31 ed                	xor    %ebp,%ebp
    10e6:	49 89 d1             	mov    %rdx,%r9
    10e9:	5e                   	pop    %rsi
    10ea:	48 89 e2             	mov    %rsp,%rdx
    10ed:	48 83 e4 f0          	and    $0xfffffffffffffff0,%rsp
    10f1:	50                   	push   %rax
    10f2:	54                   	push   %rsp
    10f3:	45 31 c0             	xor    %r8d,%r8d
    10f6:	31 c9                	xor    %ecx,%ecx
    10f8:	48 8d 3d ca 00 00 00 	lea    0xca(%rip),%rdi        # 11c9 <main>
    10ff:	ff 15 bb 2e 00 00    	call   *0x2ebb(%rip)        # 3fc0 <__libc_start_main@GLIBC_2.34>
    1105:	f4                   	hlt    
    1106:	66 2e 0f 1f 84 00 00 	cs nopw 0x0(%rax,%rax,1)
    110d:	00 00 00 

0000000000001110 <deregister_tm_clones>:
    1110:	48 8d 3d f9 2e 00 00 	lea    0x2ef9(%rip),%rdi        # 4010 <__TMC_END__>
    1117:	48 8d 05 f2 2e 00 00 	lea    0x2ef2(%rip),%rax        # 4010 <__TMC_END__>
    111e:	48 39 f8             	cmp    %rdi,%rax
    1121:	74 15                	je     1138 <deregister_tm_clones+0x28>
    1123:	48 8b 05 b6 2e 00 00 	mov    0x2eb6(%rip),%rax        # 3fe0 <_ITM_deregisterTMCloneTable@Base>
    112a:	48 85 c0             	test   %rax,%rax
    112d:	74 09                	je     1138 <deregister_tm_clones+0x28>
    112f:	ff e0                	jmp    *%rax
    1131:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)
    1138:	c3                   	ret    
    1139:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)

0000000000001140 <register_tm_clones>:
    1140:	48 8d 3d c9 2e 00 00 	lea    0x2ec9(%rip),%rdi        # 4010 <__TMC_END__>
    1147:	48 8d 35 c2 2e 00 00 	lea    0x2ec2(%rip),%rsi        # 4010 <__TMC_END__>
    114e:	48 29 fe             	sub    %rdi,%rsi
    1151:	48 89 f0             	mov    %rsi,%rax
    1154:	48 c1 ee 3f          	shr    $0x3f,%rsi
    1158:	48 c1 f8 03          	sar    $0x3,%rax
    115c:	48 01 c6             	add    %rax,%rsi
    115f:	48 d1 fe             	sar    %rsi
    1162:	74 14                	je     1178 <register_tm_clones+0x38>
    1164:	48 8b 05 85 2e 00 00 	mov    0x2e85(%rip),%rax        # 3ff0 <_ITM_registerTMCloneTable@Base>
    116b:	48 85 c0             	test   %rax,%rax
    116e:	74 08                	je     1178 <register_tm_clones+0x38>
    1170:	ff e0                	jmp    *%rax
    1172:	66 0f 1f 44 00 00    	nopw   0x0(%rax,%rax,1)
    1178:	c3                   	ret    
    1179:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)

0000000000001180 <__do_global_dtors_aux>:
    1180:	f3 0f 1e fa          	endbr64 
    1184:	80 3d 85 2e 00 00 00 	cmpb   $0x0,0x2e85(%rip)        # 4010 <__TMC_END__>
    118b:	75 2b                	jne    11b8 <__do_global_dtors_aux+0x38>
    118d:	55                   	push   %rbp
    118e:	48 83 3d 22 2e 00 00 	cmpq   $0x0,0x2e22(%rip)        # 3fb8 <__cxa_finalize@GLIBC_2.2.5>
    1195:	00 
    1196:	48 89 e5             	mov    %rsp,%rbp
    1199:	74 0c                	je     11a7 <__do_global_dtors_aux+0x27>
    119b:	48 8b 3d 66 2e 00 00 	mov    0x2e66(%rip),%rdi        # 4008 <__dso_handle>
    11a2:	e8 d9 fe ff ff       	call   1080 <__cxa_finalize@plt>
    11a7:	e8 64 ff ff ff       	call   1110 <deregister_tm_clones>
    11ac:	c6 05 5d 2e 00 00 01 	movb   $0x1,0x2e5d(%rip)        # 4010 <__TMC_END__>
    11b3:	5d                   	pop    %rbp
    11b4:	c3                   	ret    
    11b5:	0f 1f 00             	nopl   (%rax)
    11b8:	c3                   	ret    
    11b9:	0f 1f 80 00 00 00 00 	nopl   0x0(%rax)

00000000000011c0 <frame_dummy>:
    11c0:	f3 0f 1e fa          	endbr64 
    11c4:	e9 77 ff ff ff       	jmp    1140 <register_tm_clones>

00000000000011c9 <main>:
    11c9:	f3 0f 1e fa          	endbr64 
    11cd:	55                   	push   %rbp
    11ce:	48 89 e5             	mov    %rsp,%rbp
    11d1:	48 83 ec 10          	sub    $0x10,%rsp
    11d5:	89 7d fc             	mov    %edi,-0x4(%rbp)
    11d8:	48 89 75 f0          	mov    %rsi,-0x10(%rbp)
    11dc:	bf 13 00 00 00       	mov    $0x13,%edi
    11e1:	e8 ba fe ff ff       	call   10a0 <raise@plt>
    11e6:	48 8b 05 db 2d 00 00 	mov    0x2ddb(%rip),%rax        # 3fc8 <x@Base>
    11ed:	8b 00                	mov    (%rax),%eax
    11ef:	89 c6                	mov    %eax,%esi
    11f1:	48 8b 05 d8 2d 00 00 	mov    0x2dd8(%rip),%rax        # 3fd0 <_ZSt4cout@GLIBCXX_3.4>
    11f8:	48 89 c7             	mov    %rax,%rdi
    11fb:	e8 d0 fe ff ff       	call   10d0 <_ZNSolsEi@plt>
    1200:	48 8b 05 d1 2d 00 00 	mov    0x2dd1(%rip),%rax        # 3fd8 <z@Base>
    1207:	8b 00                	mov    (%rax),%eax
    1209:	89 c6                	mov    %eax,%esi
    120b:	48 8b 05 be 2d 00 00 	mov    0x2dbe(%rip),%rax        # 3fd0 <_ZSt4cout@GLIBCXX_3.4>
    1212:	48 89 c7             	mov    %rax,%rdi
    1215:	e8 b6 fe ff ff       	call   10d0 <_ZNSolsEi@plt>
    121a:	e8 91 fe ff ff       	call   10b0 <_Z3foov@plt>
    121f:	b8 00 00 00 00       	mov    $0x0,%eax
    1224:	c9                   	leave  
    1225:	c3                   	ret    

0000000000001226 <_Z41__static_initialization_and_destruction_0ii>:
    1226:	f3 0f 1e fa          	endbr64 
    122a:	55                   	push   %rbp
    122b:	48 89 e5             	mov    %rsp,%rbp
    122e:	48 83 ec 10          	sub    $0x10,%rsp
    1232:	89 7d fc             	mov    %edi,-0x4(%rbp)
    1235:	89 75 f8             	mov    %esi,-0x8(%rbp)
    1238:	83 7d fc 01          	cmpl   $0x1,-0x4(%rbp)
    123c:	75 3b                	jne    1279 <_Z41__static_initialization_and_destruction_0ii+0x53>
    123e:	81 7d f8 ff ff 00 00 	cmpl   $0xffff,-0x8(%rbp)
    1245:	75 32                	jne    1279 <_Z41__static_initialization_and_destruction_0ii+0x53>
    1247:	48 8d 05 c3 2d 00 00 	lea    0x2dc3(%rip),%rax        # 4011 <_ZStL8__ioinit>
    124e:	48 89 c7             	mov    %rax,%rdi
    1251:	e8 6a fe ff ff       	call   10c0 <_ZNSt8ios_base4InitC1Ev@plt>
    1256:	48 8d 05 ab 2d 00 00 	lea    0x2dab(%rip),%rax        # 4008 <__dso_handle>
    125d:	48 89 c2             	mov    %rax,%rdx
    1260:	48 8d 05 aa 2d 00 00 	lea    0x2daa(%rip),%rax        # 4011 <_ZStL8__ioinit>
    1267:	48 89 c6             	mov    %rax,%rsi
    126a:	48 8b 05 87 2d 00 00 	mov    0x2d87(%rip),%rax        # 3ff8 <_ZNSt8ios_base4InitD1Ev@GLIBCXX_3.4>
    1271:	48 89 c7             	mov    %rax,%rdi
    1274:	e8 17 fe ff ff       	call   1090 <__cxa_atexit@plt>
    1279:	90                   	nop
    127a:	c9                   	leave  
    127b:	c3                   	ret    

000000000000127c <_GLOBAL__sub_I_main.cpp>:
    127c:	f3 0f 1e fa          	endbr64 
    1280:	55                   	push   %rbp
    1281:	48 89 e5             	mov    %rsp,%rbp
    1284:	be ff ff 00 00       	mov    $0xffff,%esi
    1289:	bf 01 00 00 00       	mov    $0x1,%edi
    128e:	e8 93 ff ff ff       	call   1226 <_Z41__static_initialization_and_destruction_0ii>
    1293:	5d                   	pop    %rbp
    1294:	c3                   	ret    

Disassembly of section .fini:

0000000000001298 <_fini>:
    1298:	f3 0f 1e fa          	endbr64 
    129c:	48 83 ec 08          	sub    $0x8,%rsp
    12a0:	48 83 c4 08          	add    $0x8,%rsp
    12a4:	c3                   	ret    
